### PR TITLE
Expose the ChannelHandlerContext from SniHandler's select() step to t…

### PR DESCRIPTION
**Pull request by rkapsi**
…he user.

Motivation

I'm looking to harden our SSL impl. a little bit and add some guards agaist certain types of abuse. One can think of invalid hostname strings in the SNI extenstion or invalid SNI handshakes altogether. This will require measuring, velocity tracking and other things.

Modifications

Adding a protected `lookup(ctx, hostname)` method that is called from SniHandler's `select(...)` method which users can override and implement custom behaviour. The default implementation will simply call the AsyncMapper.

Result

It's possible to get a hold onto the ChannelHandlerContext. Users can override that method and do something with it right there or they can delegate it to something else. SniHandler is happy as long as a `Future<SslContext>` is being returned.

### CODE REVIEW

File: handler/src/main/java/io/netty/handler/ssl/SniHandler.java [hunk 1]
```diff
@@ -306,6 +306,16 @@ public void operationComplete(Future<SslContext> future) throws Exception {
     }
 
     /**
+     * The default implementation will simply call {@link AsyncMapping#map(Object, Promise)} but
+     * users can override this method to implement custom behavior.
+     *
+     * @see AsyncMapping#map(Object, Promise)
+     */
+    protected Future<SslContext> lookup(ChannelHandlerContext ctx, String hostname) throws Exception {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: Should we also pass the `mapping` in ? Or make it protected at least ?

Comment (rkapsi) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: Either works, do you have a preference? I think `protected` is better or maybe a getter method for it.

Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: let us just do protected for now.

File: handler/src/main/java/io/netty/handler/ssl/SniHandler.java [hunk 2]
```diff
@@ -57,7 +57,7 @@
             InternalLoggerFactory.getInstance(SniHandler.class);
     private static final Selection EMPTY_SELECTION = new Selection(null, null);
 
-    private final AsyncMapping<String, SslContext> mapping;
+    protected final AsyncMapping<String, SslContext> mapping;
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: does this need to be `protected`? can we just force users to use `super.lookup`?

File: handler/src/main/java/io/netty/handler/ssl/SniHandler.java (hunk 1 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: I should have read this before I made https://github.com/netty/netty/pull/5786#discussion_r77205752, but it seems unnecessary to expose the `mapping` member variable when users can just call `super.lookup`.

Comment (rkapsi) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: Having it `protected` could be useful for something like this:

```java
public class MyAsyncMapping implements AsyncMapping<String, SslContext> {

  @Override
  public Future<SslContext> map(String input, Promise<SslContext> promise) {
    throw new UnsupportedOperationException();
  }
  
  public Future<SslContext> lookup(ChannelHandlerContext ctx, String hostname) {
    // ... do something ...
  }
}

public class MySniHandler extends SniHandler {
  
  public MySniHandler(AsyncMapping<String, SslContext> mapper) {
    super(mapper);
  }
  
  @Override
  public Future<SslContext> lookup(ChannelHandlerContext ctx, String hostname) {
    return ((MyAsyncMapping)mapping).lookup(ctx, hostname);
  }
}
```

On the other hand, the user has the opportunity to get a hold onto the Mapper via ctor if they need to. I think having it protected in this case would save one field.

Comment (rkapsi) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: @normanmaurer @Scottmitch please let me know if you want to make changes to it.

On my end I don't need the mapper to be exposed as I'd pull it via ctor if necessary but the above code example is the most likely path I'll go in terms of using it.

Also, please take a look at #5787 if you haven't already.

Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: I think protected is fine... @Scottmitch WDYT ?

**Comment by normanmaurer** [comment]
@Scottmitch let us just make it protected for now... Let me pull this in.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (b604a22395f64bca52d155127e2453aa3178af6c)

**Comment by Scottmitch** [comment]
sorry for delay ... lgtm

**Pull request closed without merging**